blackbird_python/blackbird/tests/test_listener.py

```diff
@@ -344,6 +329,40 @@ def test_operation_arg_expressions(self, parse_input_mocked_metadata):
 
         assert bb.operations == [{"modes": [0], "op": "Coherent", "args": expected, "kwargs": {}}]
 
+    def test_operation_mode_expressions(self, parse_input_mocked_metadata):
+        """Test that expressions inside modes are properly evaluated"""
+        bb = parse_input_mocked_metadata(
+            "int m = 4\nMZgate(0, 1) | [m*2, -1+m]"
+        )
+        assert bb.operations == [{'op': 'MZgate', 'args': [0, 1], 'kwargs': {}, 'modes': [8, 3]}]
+
+    def test_parameter_idx(self, parse_input_mocked_metadata):
+        """Test that parameter array indexing works inside arguments"""
+        bb = parse_input_mocked_metadata(
+            "par array phases =\n\t{phase_0}\n\t{phase_1}\n\nMZgate(phases[0], phases[1]) | [0, 1]"
+        )
+        assert bb.operations == [
+            {'op': 'MZgate', 'args': [sym.Symbol("phase_0"), sym.Symbol("phase_1")], 'kwargs': {}, 'modes': [0, 1]}
+        ]
+
+    def test_var_idx_in_modes(self, parse_input_mocked_metadata):
+        """Test that array indexing works inside modes"""
+        bb = parse_input_mocked_metadata(
+            "int array vars =\n\t1, 2\n\nMZgate(0, 1) | [vars[0], vars[1]]"
+        )
+        assert bb.operations == [
+            {'op': 'MZgate', 'args': [0, 1], 'kwargs': {}, 'modes': [1, 2]}
+        ]
```

Comment:
This test tests the case
```python
int array vars =
    1, 2
```

Is it worth also testing the case
```python
int array vars =
    1
    2
```
(does this affect how you index?)

What about something like

```python
int array vars =
    1, 2
    3, 4
```

Reply:
I deliberately made the arrays always behave as flat, in the sense that you can write them (and store them) as nested numpy-arrays but they will behave as flat with regards to indexing. So your examples above would basically behave the same only with the last one having two more elements.

Reply:
I thought it looked nice to be able to list parameters horizontally rather than vertically,

```python
par array phases =
    {phase_0}
    {phase_1}
    {phase_2}
    {phase_3}
    {phase_4}
```

and vice versa with regards to values.

```python
int array vars =
    1, 2, 3, 4, 5, 42, 6
```

Reply:
:+1: